### DISCUSSION TITLE
Use start_server() with websocket?

### DISCUSSION THREAD

**Discussion by quintesse** [micropython/micropython#9158] [Libraries & Drivers] [answered]
I was wondering if it's possible to set up a websocket connection with a server that has been started using `start_server`?

Code examples explaining how to use `websocket` seem to show that it requires access to a socket which `start_server()` doesn't seem to provide. Is that correct or am I missing something?

**Comment by karfas** [comment]
I assume you are talking about `asyncio.start_server()` ?
This starts a simple TCP server, nothing else. Any connection made from a client needs to be handled by a callback the user provides. This callback is responsible for handling higher level protocols like HTTP (it might e.g. start an asyncio task to handle a request).

Second, I am pretty sure you need at least a minimal HTTP server capable of "promoting" a HTTP connection to a websocket connection. 
See e.g. the "Protocol Handshake" section in https://en.wikipedia.org/wiki/WebSocket. I have no idea whether one of the micropython HTTP servers supports this.

**Comment by marcidy** [comment, marked as answer]
I have an example here: https://github.com/marcidy/micropython-uasyncio-websockets with examples of use here: https://github.com/marcidy/micropython-uasyncio-webexample.  

websockets.server.serve uses uasyncio.start_server.  You'll need the protocol to be asyncio compatible as well, which that repo provides.

**Reply by quintesse** [reply]
Thanks that's very useful!
